core/matrix/sellp.hpp

```diff
@@ -0,0 +1,327 @@
+    }
+
+protected:
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *    (The total_cols is set to be the number of slice times the number
+     *     of cols of the matrix.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size = dim{})
+        : Sellp(std::move(exec), size,
+                ceildiv(size.num_rows, default_slice_size) * size.num_cols)
+    {}
+
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *    (The slice_size and stride_factor are set to the default values.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param total_cols   number of the sum of all cols in every slice.
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size,
+          size_type total_cols)
+        : Sellp(std::move(exec), size, default_slice_size,
+                default_stride_factor, total_cols)
+    {}
+
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param slice_size  number of rows in each slice
+     * @param stride_factor  factor for the stride in each slice (strides
+     *                        should be multiples of the stride_factor)
+     * @param total_cols   number of the sum of all cols in every slice.
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size,
+          size_type slice_size, size_type stride_factor, size_type total_cols)
+        : EnableLinOp<Sellp>(exec, size),
+          values_(exec, slice_size * total_cols),
+          col_idxs_(exec, slice_size * total_cols),
+          slice_lengths_(exec, (size.num_rows == 0)
+                                   ? 0
+                                   : ceildiv(size.num_rows, slice_size)),
+          slice_sets_(exec, (size.num_rows == 0)
+                                ? 0
+                                : ceildiv(size.num_rows, slice_size) + 1),
+          slice_size_(slice_size),
+          stride_factor_(stride_factor),
+          total_cols_(total_cols)
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    size_type linearize_index(size_type row, size_type slice_set,
+                              size_type col) const noexcept
+    {
+        return (slice_set + col) * slice_size_ + row;
+    }
+
+private:
+    Array<value_type> values_;
+    Array<index_type> col_idxs_;
+    Array<size_type> slice_lengths_;
+    Array<size_type> slice_sets_;
+    size_type slice_size_;
+    size_type stride_factor_;
+    size_type total_cols_;
```

Comment:
Maybe you could use `{}` to explicitely initialize some of these to 0 by default.